distributed/scheduler.py

```diff
@@ -877,363 +861,265 @@ def _to_dict_no_nest(self, *, exclude: Container[str] = ()) -> dict:
-        "traceback_text",
-        "erred_on",
-        "exception_blame",
-        "suspicious",
-        "retries",
-        "nbytes",
-        "type",
-        "group_key",
-        "group",
-        "metadata",
-        "annotations",
-    )
+    __slots__ = tuple(__annotations__)  # type: ignore
 
     def __init__(self, key: str, run_spec: object):
         self.key = key
         self._hash = hash(key)
         self.run_spec = run_spec
-        self._state: str = None  # type: ignore
-        self.exception: str = None  # type: ignore
-        self.exception_blame: TaskState = None  # type: ignore
+        self._state = None  # type: ignore
+        self.exception = None
+        self.exception_blame = None
         self.traceback = None
         self.exception_text = ""
         self.traceback_text = ""
         self.suspicious = 0
         self.retries = 0
         self.nbytes = -1
         self.priority = None  # type: ignore
-        self.who_wants: set[ClientState] = set()
-        self.dependencies: set[TaskState] = set()
-        self.dependents: set[TaskState] = set()
-        self.waiting_on: set[TaskState] = set()
-        self.waiters: set[TaskState] = set()
-        self.who_has: set[WorkerState] = set()
-        self.processing_on: WorkerState = None  # type: ignore
+        self.who_wants = set()
+        self.dependencies = set()
+        self.dependents = set()
+        self.waiting_on = set()
+        self.waiters = set()
+        self.who_has = set()
+        self.processing_on = None
         self.has_lost_dependencies = False
         self.host_restrictions = None  # type: ignore
         self.worker_restrictions = None  # type: ignore
         self.resource_restrictions = None  # type: ignore
         self.loose_restrictions = False
         self.actor = False
-        self.prefix: TaskPrefix = None  # type: ignore
-        self.type: str = None  # type: ignore
+        self.prefix = None  # type: ignore
+        self.type = None  # type: ignore
         self.group_key = key_split_group(key)
-        self.group: TaskGroup = None  # type: ignore
-        self.metadata = {}  # type: ignore
-        self.annotations = {}  # type: ignore
-        self.erred_on: set[str] = set()
+        self.group = None  # type: ignore
+        self.metadata = {}
+        self.annotations = {}
+        self.erred_on = set()
 
-    def __hash__(self):
+    def __hash__(self) -> int:
         return self._hash
 
-    def __eq__(self, other):
-        typ_self: type = type(self)
-        typ_other: type = type(other)
-        if typ_self == typ_other:
-            other_ts: TaskState = other
-            return self.key == other_ts.key
-        else:
-            return False
+    def __eq__(self, other: object) -> bool:
+        return isinstance(other, TaskState) and self.key == other.key
 
     @property
     def state(self) -> str:
+        """This task's current state.  Valid states include ``released``, ``waiting``,
```

Comment:
Thanks for this docstring! Without it, it has been extremely hard for me to understand what possible states a `TaskState` could be in.

Reply:
I'm going to replace str with a TaskStateState Literal in a later PR, like it's already done in worker_state_machine.py